docsite/rst/developing_modules.rst

```diff
@@ -23,6 +23,51 @@ repo for more established and widely used modules.  "Extras" modules may be prom
 but there's no fundamental difference in the end - both ship with ansible, all in one package, regardless
 of how you acquire ansible.
 
+Ansible Module Guidelines
+`````````````````````````
+
+These are the guidelines for people who want to contribute to the development of modules to Ansible on GitHub. Please read the guidelines before you submit your PR (proposal?).
+
+Documentation
++++++++++++++
+
+Module documentation should briefly and accurately define what each module and option does, and how it works with others in the underlying system. Documentation should be written for broad audience--readable both by experts and non-experts. This documentation is not meant to teach a total novice, but it also should not be reserved for the Illuminati (hard balance).
+
+Document the return. Refer to :ref:`common_return_values` and :ref:`module_documenting` for additional information.
+
+Predictable user interface 
++++++++++++++++++++++++++++
+
+This is a particularly important section as it is also an area where we need significant improvements.
+
+- Name consistency across modules (we’ve gotten better at this, but we still have many deviations).
+- Declarative operation (not CRUD)--this makes it easy for a user not to care what the existing state is, just about the final state. ``started/stopped``, ``present/absent``--don't overload options too much. It is preferable to add a new, simple option than to add choices/states that don't fit with existing ones.
+- Keep options small, having them take large data structures might save us a few tasks, but adds a complex requirement that we cannot easily validate before passing on to the module.
+- Allow an "expert mode". This may sound like the absolute opposite of the previous one, but it is always best to let expert users deal with complex data. This requires different modules in some cases, so that you end up having one (1) expert module and several 'piecemeal' ones (ec2_vpc_net?). The reason for this is not, as many users express, because it allows a single task and keeps plays small (which just moves the data complexity into vars files, leaving you with a slightly different structure in another YAML file). It does, however, allow for a more 'atomic' operation against the underlying APIs and services.
+
+
+Informative responses
+++++++++++++++++++++++
+
+Please note, that for  >= 2.0, it is required that return data to be documented.
+
+- Always return useful data, even when there is no change.
+- Be consistent about returns (some modules are too random), unless it is detrimental to the state/action.
+- Make returns reusable--most of the time you don't want to read it, but you do want to process it and re-purpose it.
+- return diff is in diff mode?
```

Comment:
I also think that should have read "return diff IF in diff mode."

Timothy Appnel
Principal Architect
Ansible by Red Hat

GitHub: tima <https://www.github.com/tima>

Twitter: @appnelgroup <https://www.twitter.com/@appnelgroup>

On Tue, Apr 5, 2016 at 12:37 PM, Sandra Wills <notifications@github.com>
wrote:

> In docsite/rst/developing_modules.rst
> <https://github.com/ansible/ansible/pull/15274#discussion_r58571464>:
>
> > +
> > +- Name consistency across modules (we’ve gotten better at this, but we still have many deviations).
> > +- Declarative operation (not CRUD)--this makes it easy for a user not to care what the existing state is, just about the final state. ``started/stopped``, ``present/absent``--don't overload options too much. It is preferable to add a new, simple option than to add choices/states that don't fit with existing ones.
> > +- Keep options small, having them take large data structures might save us a few tasks, but adds a complex requirement that we cannot easily validate before passing on to the module.
> > +- Allow an "expert mode". This may sound like the absolute opposite of the previous one, but it is always best to let expert users deal with complex data. This requires different modules in some cases, so that you end up having one (1) expert module and several 'piecemeal' ones (ec2_vpc_net?). The reason for this is not, as many users express, because it allows a single task and keeps plays small (which just moves the data complexity into vars files, leaving you with a slightly different structure in another YAML file). It does, however, allow for a more 'atomic' operation against the underlying APIs and services.
> > +
> > +
> > +Informative responses
> > +++++++++++++++++++++++
> > +
> > +Please note, that for  >= 2.0, it is required that return data to be documented.
> > +
> > +- Always return useful data, even when there is no change.
> > +- Be consistent about returns (some modules are too random), unless it is detrimental to the state/action.
> > +- Make returns reusable--most of the time you don't want to read it, but you do want to process it and re-purpose it.
> > +- return diff is in diff mode?
>
> The Google doc has this comment in regard to "return diff is in diff mode":
>
> Timothy Appnel/ 10:47 AM Mar 17 -- Thinking we'd want this. If I'm correct
> we should provide a bit of guidance on what to return here also.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ansible/ansible/pull/15274/files/f037e1d2ec23834c35e9a863488a821a87c84e64#r58571464>
>